### PR TITLE
ClusterLoader - Adding measurement manager

### PR DESCRIPTION
**Pull request by krzysied**
Implementation of measurement factory and measurement manager.
CloudProvider will be add in future PR.

### CODE REVIEW

File: clusterloader2/pkg/measurement/measurementmanager.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/*
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/measurementmanager.go]: Since you're already inside measurement/ directory, I think you can simply name this file `manager.go` or sth.

File: clusterloader2/pkg/measurement/measurementfactory.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/measurementfactory.go]: Since you're already inside measurement/ directory, I think you can simply name this file `factory.go` or sth.

File: clusterloader2/pkg/measurement/measurementmanager.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	lock         sync.Mutex
+	measurements map[string]map[string]Measurement
+	clientSet    clientset.Interface
+}
+
+// CreateMeasurementManager creates new instance of MeasurementManager.
+func CreateMeasurementManager(clientSet clientset.Interface) *MeasurementManager {
+	return &MeasurementManager{
+		measurements: make(map[string]map[string]Measurement),
+		clientSet:    clientSet,
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/measurementmanager.go]: Move clientset above.

File: clusterloader2/pkg/measurement/measurementmanager.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	"sync"
+
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// MeasurementManager manages all measurements calls and executions.
+type MeasurementManager struct {
+	lock         sync.Mutex
+	measurements map[string]map[string]Measurement
+	clientSet    clientset.Interface
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/measurementmanager.go]: Move clientset above.

Comment (shyamjvs) [reply clusterloader2/pkg/measurement/measurementmanager.go]: Also I don't think it needs to be protected by the lock. It's only set once during initialization.

File: clusterloader2/pkg/measurement/measurementmanager.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import (
+	"sync"
+
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// MeasurementManager manages all measurements calls and executions.
+type MeasurementManager struct {
+	lock         sync.Mutex
+	measurements map[string]map[string]Measurement
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/measurementmanager.go]: Please add a comment above this complex map explaining what it's storing.

File: clusterloader2/pkg/measurement/measurementfactory.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+*/
+
+package measurement
+
+import (
+	"fmt"
+	"sync"
+)
+
+var factory *measurmentsFactory
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/measurementfactory.go]: spelling mistake

Comment (shyamjvs) [reply clusterloader2/pkg/measurement/measurementfactory.go]: Also change it to `measurementFactory` (i.e plural isn't needed).

File: clusterloader2/pkg/measurement/measurementmanager.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+		return err
+	}
+	config := &MeasurementConfig{
+		ClientSet: mm.clientSet,
+		Params:    params,
+	}
+	return m.Execute(config)
+}
+
+func (mm *MeasurementManager) getMeasurement(identifier string, metricMethod string) (Measurement, error) {
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/measurementmanager.go]: Why are you calling it metricMethod and not just method? (because it doesn't have to be a metric-related method)

File: clusterloader2/pkg/measurement/measurementmanager.go (hunk 1 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/measurementmanager.go]: done

File: clusterloader2/pkg/measurement/measurementfactory.go (hunk 2 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/measurementfactory.go]: done

File: clusterloader2/pkg/measurement/measurementmanager.go (hunk 3 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/measurementmanager.go]: done

File: clusterloader2/pkg/measurement/measurementmanager.go (hunk 4 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/measurementmanager.go]: done

File: clusterloader2/pkg/measurement/measurementmanager.go (hunk 5 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/measurementmanager.go]: done

File: clusterloader2/pkg/measurement/measurementfactory.go (hunk 6 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/measurementfactory.go]: done

File: clusterloader2/pkg/measurement/manager.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	"sync"
+
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// MeasurementManager manages all measurements calls and executions.
+type MeasurementManager struct {
+	clientSet    clientset.Interface
+	lock         sync.Mutex
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/manager.go]: Please add newline above lock.

File: clusterloader2/pkg/measurement/manager.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+package measurement
+
+import (
+	"sync"
+
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// MeasurementManager manages all measurements calls and executions.
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/manager.go]: -> measurement

Comment (shyamjvs) [reply clusterloader2/pkg/measurement/manager.go]: Also just executions is sufficient I guess.

File: clusterloader2/pkg/measurement/manager.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+type MeasurementManager struct {
+	clientSet    clientset.Interface
+	lock         sync.Mutex
+	// measurements is a map: methodName -> identifier -> Measurement.
+	// First level divides measurements into method classes, second into instances.
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/manager.go]: I think this comment is a bit redundant. Maybe simply say 'Map from method type and identifier to measurement instance'?

File: clusterloader2/pkg/measurement/factory.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+package measurement
+
+import (
+	"fmt"
+	"sync"
+)
+
+var factory *measurementFactory
+
+func init() {
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/factory.go]: Why do you need to have global instance? Like I mentioned earlier, it makes it hard to have 2 instances of this pkg.

File: clusterloader2/pkg/measurement/interface.go [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+package measurement
+
+import (
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// MeasurementConfig provides client and paramters required for the measurement execution.
+type MeasurementConfig struct {
+	// Clientset is a kubernetes cloud client.
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/interface.go]: Remove 'cloud'

File: clusterloader2/pkg/measurement/interface.go [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+	ClientSet clientset.Interface
+	// Params is a map {name: value} pairs enabling for injection of arbitrary config
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/interface.go]: map of

File: clusterloader2/pkg/measurement/interface.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	// TODO(krzysied): add CloudProvider.
+}
+
+// Measurement interface
+type Measurement interface {
+	Execute(config *MeasurementConfig) error
+}
+
+// CreateMeasurementFunc is a type of create function used by the measurement factory.
+type CreateMeasurementFunc func() Measurement
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/interface.go]: I don't think this type needs to be public.

File: clusterloader2/pkg/measurement/manager.go [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+		return err
+	}
+	config := &MeasurementConfig{
+		ClientSet: mm.clientSet,
+		Params:    params,
+	}
+	return m.Execute(config)
+}
+
+func (mm *MeasurementManager) getMeasurement(identifier string, methodName string) (Measurement, error) {
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/manager.go]: -> getMeasurementObject? (because getMeasurement sounds ambiguous - it could even mean getting a real measurement)

Comment (shyamjvs) [reply clusterloader2/pkg/measurement/manager.go]: Nit - could you reverse the argument order of the function? (to match key order in the map)

File: clusterloader2/pkg/measurement/manager.go (hunk 8 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/manager.go]: done

File: clusterloader2/pkg/measurement/manager.go (hunk 9 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/manager.go]: done

File: clusterloader2/pkg/measurement/manager.go (hunk 10 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/manager.go]: done

File: clusterloader2/pkg/measurement/factory.go (hunk 11 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/factory.go]: I've added function that can create new instance of MeasurementFactory. Though, I'm still keeping global factory as default instance.

File: clusterloader2/pkg/measurement/interface.go (hunk 12 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/interface.go]: done

File: clusterloader2/pkg/measurement/interface.go (hunk 13 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/interface.go]: done

File: clusterloader2/pkg/measurement/manager.go (hunk 15 above)
Comment (krzysied) [reply clusterloader2/pkg/measurement/manager.go]: `getMeasurementInstance` it is.
Order of parameters changed.

File: clusterloader2/pkg/measurement/manager.go [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+	lock sync.Mutex
+	// measurements is a map from method type and identifier to measurement instance.
+	measurements map[string]map[string]Measurement
+}
+
+// CreateMeasurementManager creates new instance of MeasurementManager.
+func CreateMeasurementManager(factory *MeasurementFactory, clientSet clientset.Interface) *MeasurementManager {
+	return &MeasurementManager{
+		factory:      factory,
```
[the remark this answers is not in the document]
Comment (shyamjvs) [reply clusterloader2/pkg/measurement/manager.go]: This way users don't need to worry about managing 2 objects (i.e both factory and manager), but only the latter.

File: clusterloader2/pkg/measurement/manager.go [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+		clientSet:    clientSet,
+		measurements: make(map[string]map[string]Measurement),
+	}
+}
+
+// Execute executes measurement based on provided identifier, methodName and params.
+func (mm *MeasurementManager) Execute(methodName string, identifier string, params map[string]interface{}) error {
+	mInstance, err := mm.getMeasurementInstance(methodName, identifier)
```
Comment (shyamjvs) [inline clusterloader2/pkg/measurement/manager.go]: You can as well name it `measurementInstance` instead of `mInstance` :)

File: clusterloader2/pkg/measurement/manager.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (shyamjvs) [reply clusterloader2/pkg/measurement/manager.go]: Infact, you could just merge factory and manager into single struct, as each of them pretty much holds 1 variable.

[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/pkg/measurement/manager.go]: > Infact, you could just merge factory and manager into single struct, as each of them pretty much holds 1 variable.

TL;DR: I will be very unhappy to merge those two structures.
Those two are two separated concept. Factory should be something on global level, that is aware of how to create measurement instances.
Manager exists in some context (test context in this case) - it uses factory to create measurement instances. Managers also stores states of instances.

> Initialize a new instance of factory in CreateMeasurementManager (so you don't need that global variable anymore)
Make Register() a public method on the MeasurementManager and make the Register() for the factory a private one, and call the latter from the former

Could you explain to me, how will measurement methods will register in factory? I still missing the point where Register is called by the measurement method implementation or measurement is added to manager.

With this implementation I try to do similar thing to how apis register to legacyscheme.

[the remark this answers is not in the document]
Comment (shyamjvs) [reply clusterloader2/pkg/measurement/manager.go]: If factory is something that is really independent of manager, why are you holding it as a field in the manager? So I think it makes sense to either:

- Keep the factory independent of manager and hold them separately, or
- Make manager manage everything. By this I don't mean merging both structs (that was just another suggestion), but I meant that from the cluster-loader perspective it only talks to manager

> Could you explain to me, how will measurement methods will register in factory?

How would it be done currently? It'll be just the same, but instead of calling Register() on factory, you'll call it on manager which will then be responsible for storing it.

[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/pkg/measurement/manager.go]: > If factory is something that is really independent of manager, why are you holding it as a field in the manager?

Factory is something independent. It's manager that depends on factory. Managers uses factory the same way it uses clientSet.
Plus, I can easily imagine having factory initialize once and then have multiple tests run with separate manager each that use this one factory. Therefore is think those two thing are separate concepts.

[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/pkg/measurement/manager.go]: > How would it be done currently? It'll be just the same, but instead of calling Register() on factory, you'll call it on manager which will then be responsible for storing it.

I'm keeping one instance of factory as global variable so measurements can register themselves. Does it mean that using your approach I should make one global manager? Or after creating manager should I explicitly register each measurement? - This part is still not clear for me :(

[the remark this answers is not in the document]
Comment (shyamjvs) [reply clusterloader2/pkg/measurement/manager.go]: - If factory is independent of manager's lifecycle, then why are your storing it as a field in the manager?
- If factory can exist without manager, why are you making createMeasurement() private? Because that's also something a user of factory would want besides Register()

> Therefore is think those two thing are separate concepts.

I agree and am not saying they're not. It's about what really do you want to expose to users of this package.

[the remark this answers is not in the document]
Comment (shyamjvs) [reply clusterloader2/pkg/measurement/manager.go]: > I'm keeping one instance of factory as global variable so measurements can register themselves. Does it mean that using your approach I should make one global manager?

Yes. The only change you should need to make is having a manager instance instead of a manager and a factory instance (whichever way you're defining them.. global or local or field ..)

[the remark this answers is not in the document]
Comment (krzysied) [reply clusterloader2/pkg/measurement/manager.go]: > If factory is independent of manager's lifecycle, then why are your storing it as a field in the manager?

Manager is using it therefore pointer is needed. In theory, someone can have separate factory instance. So there should be a way to choose which factory should be used.

TBH, I still prefer to have factory singleton in measurement module (as it was done in first place). This way there is no need to pass factory pointers etc.

> If factory can exist without manager, why are you making createMeasurement() private? Because that's also something a user of factory would want besides Register()

That's good observation. I should make createMeasurement public.

> Yes. The only change you should need to make is having a manager instance instead of a manager and a factory instance (whichever way you're defining them.. global or local or field ..)

Manager is something that should exist only on test level, thus can't be made global. On the other hand factory in independent of any test state so this one can be made global.

**Comment by shyamjvs** [comment]
/lgtm
Thanks!
